libvast/src/transform.cpp

```diff
@@ -0,0 +1,146 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2016 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "vast/transform.hpp"
+
+#include "vast/arrow_table_slice_builder.hpp"
+#include "vast/logger.hpp"
+#include "vast/table_slice_builder.hpp"
+#include "vast/table_slice_builder_factory.hpp"
+
+#include <arrow/type.h>
+
+namespace vast {
+
+caf::expected<table_slice> transform_step::apply(table_slice&& slice) const {
+  bool enable_arrow = VAST_ENABLE_ARROW > 0;
+  const auto* arrow_step = dynamic_cast<const arrow_transform_step*>(this);
+  const auto* generic_step = dynamic_cast<const generic_transform_step*>(this);
+  if (arrow_step && enable_arrow) {
+    return (*arrow_step)(std::move(slice));
+  }
```

Comment:
There should be an assert here (and when we load the transform plugin) that either arrow_step or generic_step is non-null.